org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/util/CancelIndicatorProgressMonitor.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2016, 2020 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ide.util;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.xtext.util.CancelIndicator;
+
+/**
+ * @author kosyakov - Initial contribution and API
+ * @since 2.11
+ */
+public class CancelIndicatorProgressMonitor implements IProgressMonitor {
+	private final CancelIndicator delegate;
+
+	private boolean canceled;
```

Comment:
`volatile`?

Reply:
to make sure works multithreaded?

Reply:
Yes. I'd assume that the canceled flag will be set by a thread different from the current worker thread. Otherwise the worker would just bail out.